src/modules/kvs/kvstxn.c

```diff
@@ -769,56 +769,43 @@ static int add_missing_ref (kvstxn_t *kt, const char *ref)
     return -1;
 }
 
-/* normalize key for setroot, and add it to keys array, if unique */
-static int normalize_and_append_unique (json_t *keys, const char *key)
+/* normalize key for setroot, and add it to keys dict, if unique */
+static int normalize_and_add_unique (json_t *keys, const char *key)
 {
     char *key_norm;
-    size_t index;
-    json_t *value;
-    bool unique = true;
+    int rc = -1;
 
     if ((key_norm = kvs_util_normalize_key (key, NULL)) == NULL)
         return -1;
-    json_array_foreach (keys, index, value) {
-        const char *s = json_string_value (value);
-        if (s && !strcmp (s, key_norm)) {
-            unique = false;
-            break;
-        }
-    }
-    if (unique) {
-        json_t *o;
-        if (!(o = json_string (key_norm)))
-            goto error;
-        if (json_array_append_new (keys, o) < 0) {
-            json_decref (o);
+    if (!json_object_get (keys, key_norm)) {
+        /* we don't need the value, just use a json null */
+        if (json_object_set_new (keys, key_norm, json_null ()) < 0)
```

Comment:
It would probably be OK to skip the initial test and just (re-)set.  Slightly simpler and maybe a little faster.